engine/access/ingestion/collection_syncer.go

```diff
@@ -0,0 +1,398 @@
+func (s *CollectionSyncer) requestMissingCollectionsBlocking(ctx context.Context) error {
+	collections, _, err := s.findMissingCollections(s.lastFullBlockHeight.Value())
+	if err != nil {
+		return err
+	}
+	if len(collections) == 0 {
+		s.logger.Info().Msg("skipping downloading missing collections. no missing collections found")
+		return nil
+	}
+
+	s.requestCollections(collections, true)
+
+	collectionsToBeDownloaded := make(map[flow.Identifier]struct{})
+	for _, collection := range collections {
+		collectionsToBeDownloaded[collection.CollectionID] = struct{}{}
+	}
+
+	collectionStoragePollTicker := time.NewTicker(collectionCatchupDBPollInterval)
+	defer collectionStoragePollTicker.Stop()
+
+	// we want to wait for all collections to be downloaded so we poll local storage periodically to make sure each
+	// collection was successfully saved in the storage.
+	for len(collectionsToBeDownloaded) > 0 {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("failed to complete collection retrieval: %w", ctx.Err())
+
+		case <-collectionStoragePollTicker.C:
+			s.logger.Info().
+				Int("total_missing_collections", len(collectionsToBeDownloaded)).
+				Msg("retrieving missing collections...")
+
+			for collectionID := range collectionsToBeDownloaded {
+				downloaded, err := s.isCollectionInStorage(collectionID)
+				if err != nil {
+					return err
+				}
+
+				if downloaded {
+					delete(collectionsToBeDownloaded, collectionID)
+				}
+			}
+		}
+	}
+
+	s.logger.Info().Msg("collection catchup done")
+	return nil
+}
+
+// findMissingCollections scans block heights from last known full block up to the latest finalized
+// block and returns all missing collection along with the count of incomplete blocks.
+//
+// No errors expected during normal operations.
+func (s *CollectionSyncer) findMissingCollections(lastFullBlockHeight uint64) ([]*flow.CollectionGuarantee, int, error) {
+	// first block to look up collections at
+	firstBlockHeight := lastFullBlockHeight + 1
+
+	lastFinalizedBlock, err := s.state.Final().Head()
+	if err != nil {
+		return nil, 0, err
+	}
+	// last block to look up collections at
+	lastBlockHeight := lastFinalizedBlock.Height
+
+	var missingCollections []*flow.CollectionGuarantee
+	var incompleteBlocksCount int
+
+	for currBlockHeight := firstBlockHeight; currBlockHeight <= lastBlockHeight; currBlockHeight++ {
+		collections, err := s.findMissingCollectionsAtHeight(currBlockHeight)
+		if err != nil {
+			return nil, 0, ErrMissingCollection
+		}
+
+		if len(collections) == 0 {
+			continue
+		}
+
+		missingCollections = append(missingCollections, collections...)
+		incompleteBlocksCount += 1
+	}
+
+	return missingCollections, incompleteBlocksCount, nil
+}
+
+// findMissingCollectionsAtHeight returns all missing collections for a specific block height.
+//
+// Expected errors during normal operations:
+// - storage.ErrNotFound if no block is found in storage for the given height
```

Comment:
since the block is expected to be finalized, the block MUST exist in storage. If it doesn't then it's an exception. Let's rename this to be clearer about the expectation
```suggestion
// findMissingCollectionsAtFinalizedHeight returns all missing collections for a specific finalized block height.
//
// No errors expected during normal operation
```